samples/net/sockets/http_server/src/server.c

```diff
@@ -0,0 +1,721 @@
+			    sizeof(serv_addr)) < 0) {
+			LOG_ERR("bind");
+			return -1;
+		}
+	} else if (config->address_family == AF_INET6) {
+		struct sockaddr_in6 serv_addr;
+
+		memset(&serv_addr, 0, sizeof(serv_addr));
+		serv_addr.sin6_family = AF_INET6;
+		serv_addr.sin6_addr = in6addr_any;
+		serv_addr.sin6_port = htons(config->port);
+
+		if (bind(ctx->server_fd, (struct sockaddr *)&serv_addr,
+			    sizeof(serv_addr)) < 0) {
+			LOG_ERR("bind");
+			return -1;
+		}
+	}
+
+	/* Listen for connections */
+	if (listen(ctx->server_fd, MAX_CLIENTS) < 0) {
+		LOG_ERR("listen");
+		return -1;
+	}
+
+	/* Create an eventfd*/
+	ctx->event_fd = eventfd(0, 0);
+	if (ctx->event_fd < 0) {
+		LOG_ERR("eventfd");
+		return -1;
+	}
+
+	/* Initialize client_fds */
+	memset(ctx->client_fds, 0, sizeof(ctx->client_fds));
+	memset(ctx->clients, 0, sizeof(ctx->clients));
+
+	ctx->client_fds[0].fd = ctx->server_fd;
+	ctx->client_fds[0].events = POLLIN;
+
+	ctx->client_fds[1].fd = ctx->event_fd;
+	ctx->client_fds[1].events = POLLIN;
+
+	ctx->num_clients = 0;
+	ctx->infinite = 1;
+
+	return ctx->server_fd;
+}
+
+int http2_server_start(struct http2_server_ctx *ctx)
+{
+	printf("\nType 'quit' to quit\n\n");
+	printf("Waiting for incoming connections...\n");
+
+	eventfd_t value;
+
+	do {
+		int ret = poll(ctx->client_fds, ctx->num_clients + 2, 1000);
+
+		if (ret < 0) {
+			LOG_ERR("poll failed");
+			return -1;
+		}
+
+		for (int i = 0; i <= ctx->num_clients + 1; i++) {
+			if (ctx->client_fds[i].revents & POLLERR) {
+				LOG_ERR("Error on fd %d\n",
+					ctx->client_fds[i].fd);
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			if (ctx->client_fds[i].revents & POLLHUP) {
+				LOG_INF("Client on fd %d has disconnected\n",
+					ctx->client_fds[i].fd);
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			if (!(ctx->client_fds[i].revents & POLLIN))
+				continue;
+
+			if (i == 0) {
+				int new_socket =
+					accept_new_client(ctx->server_fd);
+
+				for (int j = 1; j < MAX_CLIENTS; j++) {
+					if (ctx->client_fds[j].fd != 0)
+						continue;
+
+					ctx->client_fds[j].fd = new_socket;
+					ctx->client_fds[j].events = POLLIN;
+
+					if (j > ctx->num_clients)
+						ctx->num_clients++;
+
+					break;
+				}
+				continue;
+			}
+
+			if (i == 1) {
+				eventfd_read(ctx->event_fd, &value);
+				printf("Received stop event. exiting ..\n");
+
+				return 0;
+			}
+
+			int valread = recv(
+				ctx->client_fds[i].fd, buffer, BUFFER_SIZE, 0);
+
+			if (valread < 0) {
+				LOG_ERR("ERROR reading from socket");
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			if (valread == 0) {
+				LOG_INF("Connection closed by peer.\n");
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			/* Check the client preface */
+			if (strncmp(buffer, preface, strlen(preface)) != 0) {
+				LOG_INF("Client does not support HTTP/2.\n");
+				handle_http1_request(ctx, i, &valread);
+				continue;
+			}
+
+			handle_http2_request(ctx, i, &valread);
```

Comment:
This comment will be a little longer, but I'll try to give a general overview of how a stateful parsing could look like.

First of all, we want to get rid of all poll()/recv() calls outside of the main server loop (so the one here). The reason for this is that we don't want to block the server from handling other requests, in case `recv()` would block inside `handle_http2_request()`. But even with non-blocking calls it doesn't work great in the current design, as in case there are no new data, you'd stop the parsing and start over from scratch (i. e. preface handling) w/o serving the request. 

### Receving new data

What I think would be needed in the first place to achieve this goal is to redesign how we recv new data, I suggest to:
* Introduce a separate input buffer for each client (so make it a part of the client context).
* Introduce a new `offset` variable to keep track of how many bytes are currently in the buffer. This would also affect how you `recv()` new data:
```c
recv(ctx->client_fds[i].fd, buffer + offset, sizeof(buffer) - offset, 0);
```
As you don't want to overwrite the data that is already in the buffer.

### Parsing the buffer

Now, `recv()` will be providing new data to the buffer, so we need `handle_http_request` to consume it. That's where the stateful handling kicks in. If you look at the existing HTTP1 parser implementation you'll see there are lots of states, the parser also works in a way that it counsumes the new data byte by byte:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/http/http_parser.c#L784
I don't think we need it to get that complex with HTTP2 since it's binary, so easier to parse.

First of all, we need to figure out what states the server can be in. The very first thing we expect from a newly connected client is a preface - therfore, it makes sense to introduce a state that represents this stage (`HTTP_PREFACE`). Once we've received the preface, we can determine whether we deal with HTTP1 or HTTP2 request. 

For HTTP1 I believe we don't need much state representation, as most of this is handled by the existing parser. `HTTP1_REQUEST` should do, but if you see a need for more states for HTTP1 parsing, don't hesitate to add more.

For HTTP2 it gets a bit more complex, as we need to implement the parser ourselves. So we need to figure out, what is the next data we expect to get, when we proceed with HTTP2 request. For me, it seems that we should expect frame header. So we can introduce `HTTP2_FRAME_HEADER` state. Once we're done with the frame header parsing, we know what type of frame we deal with. So for a header frame, we could have `HTTP2_HEADER_FRAME` in which we parse header fields one by one. Similarly, in `HTTP2_SETTINGS_FRAME` we would parse ("consume") settings, one by one (note that they all have predefined format/size). The list goes on and on, I hope that you see where it's going. If certain frame type requires more complex parsing - you can define substates as well.

Also note, that with each frame header, you know exactly how much bytes in that frame to expect. So when you're done with parsing of the current frame ("consumed" all of the frame data), you can get back to `HTTP2_FRAME_HEADER` state.

### Some pseudocode to consider

We should consider now how could this look like in practice (note that it's just a pseudo-C code, I did not compile it, it's just for an overview).

The main input function could look like:
```c
int handle_http_request(struct http2_client_ctx *ctx)
{
int handle_http_request(struct http2_client_ctx *ctx)
{
	int ret = -EINVAL;

	do {
		switch (ctx->state) {
		case HTTP_PREFACE:
			ret = handle_http_preface(ctx);
			break;
		case HTTP1_REQUEST:
			ret = handle_http1_request(ctx);
			break;
		case HTTP2_FRAME_HEADER:
			ret = handle_http2_frame_header(ctx);
			break;
		case xxx:
			etc..
		}
	} while (ret == 0 && ctx->offset > 0);

	return ret;
}
```

We need to establish some protocol for the return values. 0 could indicate successful parsing and that the function should proceed with the next state (there might still be some data in the buffer to process). `-EAGAIN` could indicate that we expect more data to come before we can proceed with parsing. Any other negative value - parsing error, we should terminate the connection.

Now for example how could `handle_preface()` look like:
```c
int handle_http_preface(struct http2_client_ctx *ctx)
{
	if (ctx->offset < strlen(preface)) {
		/* We don't have full preface yet, get more data. */
		return -EAGAIN;
	}

	if (strncmp(ctx->buffer, preface, strlen(preface)) != 0) {
		ctx->state = HTTP1_REQUEST;
	} else {
		ctx->state = HTTP2_FRAME_HEADER;
	}

	/* Now we need to "consume" the already processed data from the buffer
	 * For simplicity, I suggest to use memmove for starters, we can
	 * optimize to some other solution if needed
	 */

	/* ctx->offset will indicate remaining data in the buffer */
	ctx->offset -= strlen(preface);
	memmove(ctx->buffer, ctx->buffer + strlen(preface), ctx->offset);

	return 0;
}
```